Add a `ProviderManager`

Mampfinator opened the pull request.

Mampfinator committed:
fix corrupted deno.lock file

Mampfinator committed:
initial `ProviderManager` implementation

Mampfinator committed:
initial `ProviderManager` implementation

Mampfinator committed:
Merge branch 'Tahvohck:master' into provider-manager

Tahvohck committed:
I got lazy about typing this out every time

Tahvohck committed:
I got lazy about typing this out every time

Tahvohck committed:
remove issue in lockfile

Tahvohck committed:
Fix: ProviderConfig now ACTUALLY saves

Tahvohck committed:
heheh funny

Tahvohck committed:
Remove save() changing SHOULD_SAVE

I can't think of any good reason why save() should be changing that. It edits no properties on the object and we probably shouldn't just assume it's safe to turn saving back on once complete.

Tahvohck committed:
Change how save location is passed

This way configs don't need to describe a constructor if they have no special constructor code. It also enforces a save location being passed in in the first place.

Tahvohck committed:
Implement ProviderConfig on demo provider

The demo provider should always be up to date as the gold standard for a provider.

Tahvohck committed:
Move config into you know. The config.

Tahvohck committed:
add extra comments

Tahvohck committed:
Add method for configs to validate (optional)

Tahvohck committed:
Wrap unit tests in import.meta.main block

Tahvohck committed:
Revert "Wrap unit tests in import.meta.main block"

This reverts commit 1fcee0d91bede9e2db07e8b6206d9829b31b1285.
It turns out that during testing, import.meta.main is false.

Tahvohck committed:
Add unit tests

Tahvohck committed:
Changing import for in-app files to be clearer

@/ was a bit ambiguous as to what it actually meant, so changing it to @app/ should be more clear.

Tahvohck committed:
cherry-pick aa8ad8a1e6687ce4371ec45f48fd7a88cd5528d4

Tahvohck committed:
Changing import for in-app files to be clearer

@/ was a bit ambiguous as to what it actually meant, so changing it to @app/ should be more clear.

Tahvohck committed:
make validate actually optional instead of weirdly optional

I won the coin flip :)

Mampfinator committed:
simplify `Combine` loop logic

Mampfinator committed:
add tests and minor fixes to `Combine`

Mampfinator committed:
add shared `Combine` stream to `ProviderManager`

Mampfinator committed:
`Combine` and `ProviderManager` improvements

Mampfinator committed:
Merge branch 'master' into provider-manager

Mampfinator committed:
`deno fmt` relevant files

Tahvohck committed:
Break config class out into its own file

In preparation for making ProviderConfig into a general locally saved config, there's nothing inherently linked to the providers other than that's where we initially ran into needing them.

Tahvohck committed:
Rename class to reflect the change to generic

Tahvohck committed:
Update comments and documentation

Tahvohck committed:
Fix imports

these worked fine in production but vscode thought they were relative imports

Tahvohck committed:
Skip auto validation during internal plumbing

Tahvohck committed:
Default behavior should be to save unless internally disabled.

Tahvohck committed:
Move tests into dedicated test file

Tahvohck committed:
assertRejects needs to be awaited

Tahvohck committed:
Add message to test error to avoid blank lines

Tahvohck committed:
default this back to false

After discussion this should only be true if a clean load from disk was successful.

Tahvohck committed:
Rename to be more explicit about behavior

Tahvohck committed:
break code into reusable function

Tahvohck committed:
Remove unused test

Tahvohck committed:
clean tests better at start

Tahvohck committed:
fix broken tests

Tahvohck committed:
Load files without fucking user config

Previously this was very likely to overwrite user configuration. JSON parse has been broken out from file load so that file can safely be created if not present without overwriting the file if the json fails.

Tahvohck committed:
Update test name to be clearer

Tahvohck commented on src/util.ts:
```diff
@@ -1,3 +1,127 @@
+    return out;
+};
+
+Deno.test({
```
These tests can be broken out into a standalone file without issue.
Also, the tests run pretty long, so let's slash all the delays by 10x.

Mampfinator replied:
Done. I'll slash the delays in "preserves item order" by 10x (since they were 100/350, leading to just under a second of runtime), but I won't cut them as drastically in the other one because of JS async timing shenanigans. I can get both tests below 100ms if that's fast enough for you.
![image](https://github.com/user-attachments/assets/e3c61d30-636d-4666-af20-515428c0466e)

Tahvohck replied:
Yeah, that's good enough. They were just too long at that time.

Tahvohck commented on src/util.ts:
```diff
@@ -1,3 +1,127 @@
+ * Combines multiple async iterables into one, while also allowing you to remove or add iterables during use.
+ *
+ * When any iterable is exhausted, it is automatically removed. When there are no more iterables to process. the iterator ends.
+ */
```
There's no need for a blank line in this documentation, let's get rid of that.

Tahvohck commented on src/chat_providers/YouTube.ts:
```diff
@@ -113,6 +109,13 @@ export class YouTubeDonationProvider implements DonationProvider {
+                this.config.streamId = newValue;
+            },
+        );
     }
```
I'm not sure we should be adding this during this PR. I'm not fully against it, but it still might be better to do this in a later feature branch.

Tahvohck commented on src/ProviderManager.ts:
```diff
@@ -0,0 +1,86 @@
+    constructor() {
+        super('providers.json');
+    }
+}
```
This will become an invalid instantiation once changes from #5 are pulled in. Also, now that I'm reviewing the file, it would probably be better to just include a field for `providerStatus: Record<string, boolean>` instead of this weird interface merging.

Tahvohck replied:
Alternatively, we could allow for configs to provide a remap to a different sub-variable, but that sounds like way more of a pain in the ass than it needs to be.

Mampfinator replied:
> This will become an invalid instantiation once changes from #5 are pulled in. Also, now that I'm reviewing the file, it would probably be better to just include a field for `providerStatus: Record<string, boolean>` instead of this weird interface merging.

We can't! Configs don't (yet) support recursively proxying object-type properties. I can look into that, but that might add a lot of complexity. So this interface merge is our best bet for now.

Tahvohck replied:
This is now an option with #8 merged in.

Mampfinator replied:
Done.

Tahvohck commented on src/main.ts:
```diff
@@ -1,21 +1,33 @@
-console.log('Program complete');
+await manager.activateAll();
+
+const cap = Math.ceil(Math.random() * 50);
```
this is a pretty excessive number of messages to show while we're in our testing phase. Just set it to like, 4 as a constant.

Mampfinator replied:
~~Boring but~~ sure

Tahvohck committed:
Add complex case test (broken)

Tahvohck committed:
remove ugly as sin typing

honestly I blame it being 2:30am for choosing that typing, but now it's gone

Mampfinator committed:
remove unnecessary newline from `Combine` docs

Mampfinator committed:
break `Combine` tests out into separate file, slash delays for shorter tests

Mampfinator committed:
set message cap to lower (constant) value for testing

Mampfinator committed:
remove configure logic in YouTube provider
will be added back in a future branch

Mampfinator committed:
allow `SavedConfig` to also trap and save on nested property sets
and update the complicated test accordingly. We can't restore classes on load, so that has been removed.

Mampfinator committed:
improve order of some calls/statements to reduce unnecessary work

Mampfinator committed:
guard against potentially indexing null properties

Mampfinator committed:
Revert "improve order of some calls/statements to reduce unnecessary work"

This reverts commit c74dc99d2daa2dc70aaf81a8b16d4383a081c000.

Tahvohck committed:
automatic format pass

Tahvohck committed:
mild whitespace and comment cleanup

Tahvohck committed:
BUGFIX: skip validation and saving on symbols

Mampfinator committed:
Merge branch 'master' into provider-manager

Mampfinator committed:
migrate `ProviderManagerConfig` to new config system

Mampfinator committed:
run `demo fmt`

Mampfinator committed:
Be more explicit about providers that exist in the config, but aren't registered
